tracing/src/main/java/com/palantir/tracing/TraceLocal.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+public final class TraceLocal<T> {
+
+    private final Supplier<T> initialValue;
```

Comment:
What are your expectations for the initialValue, how should it be used?
In a world with listeners, is there an expectation that listeners are invoked for a TraceLocal with an initial value which has never had `get` or `set` called within a given trace?

Reply:
yeah, maybe the potential for confusion isn't worth the convenience.

I was imagining: if get/set never called, then listeners are not invoked.

For convenience, it would be nice to have thread locals like `TraceLocal<Counter> counter = new TraceLocal(Counter::new)` which you could easily just use like `counter.get().inc()`...

Reply:
added a test to demonstrate current behaviour

Reply:
I think the initialValue component is reasonable, but should be documented to describe the edge cases

Reply:
added javadoc